rivertest/example_require_inserted_test.go

```diff
@@ -54,7 +54,13 @@ func Example_requireInserted() {
 		panic(err)
 	}
 
-	_, err = riverClient.Insert(ctx, RequiredArgs{
+	tx, err := dbPool.Begin(ctx)
+	if err != nil {
+		panic(err)
+	}
+	defer func() { _ = tx.Rollback(ctx) }()
```

Comment:
Hm, I'm not sure I totally get the concern in this one — isn't this the same deferred transaction roll back that we're already doing in ten other places?

The rollback might not happen on context cancel, but that might actually make sense. If a context were to be cancelled, it'd cancel other operations ongoing here (or elsewhere), then as the function was returning the defer would invoke roll back. If the roll back cancelled itself on context cancellation, it'd never be issued, because we're already cancelled by that point. So this allows the rollback to succeed normally, even in the event of a cancellation, therefore helping to clean up this resource.

Does that sound right to you?

Reply:
(Going to merge this since I have a docs page incoming, but will continue discussion on this thread.)

Reply:
I'm not sure why I was linking to the `Begin()` docs while talking about `Rollback` 🤦‍♂️  However I guess I'm still a bit confused how the contexts work with begin + rollback even though I thought I understood this previously. Begin specifically says the context only affects the _begin_ operation and not the rest of the transaction, whereas AFAICT Rollback does not say anything about how it behaves in response to a cancelled context. 

If my context terminates (times out) midway through my transaction, does `tx.Rollback(ctx)` do anything? How does the transaction get aborted in that case?

Reply:
Hmm, actually, by my read I think it does respect cancellation ...

You have to trace code all over the place, so I'm not entirely sure, but I think it does.

`Rollback`:

```go
func (tx *dbTx) Rollback(ctx context.Context) error {
	if tx.closed {
		return ErrTxClosed
	}

	_, err := tx.conn.Exec(ctx, "rollback")
	tx.closed = true
	if err != nil {
		// A rollback failure leaves the connection in an undefined state
		tx.conn.die(fmt.Errorf("rollback failed: %w", err))
		return err
	}

	return nil
}
```

`Exec` calls into another `exec`:

```go
func (c *Conn) Exec(ctx context.Context, sql string, arguments ...any) (pgconn.CommandTag, error) {
	if c.queryTracer != nil {
		ctx = c.queryTracer.TraceQueryStart(ctx, c, TraceQueryStartData{SQL: sql, Args: arguments})
	}

	if err := c.deallocateInvalidatedCachedStatements(ctx); err != nil {
		return pgconn.CommandTag{}, err
	}

	commandTag, err := c.exec(ctx, sql, arguments...)
```

Which goes through half a doze subfunctions, but eventually ends up somewhere like `execExtendedPrefix`:

```go
func (pgConn *PgConn) execExtendedPrefix(ctx context.Context, paramValues [][]byte) *ResultReader {
	pgConn.resultReader = ResultReader{
		pgConn: pgConn,
		ctx:    ctx,
	}
	result := &pgConn.resultReader

...

	if ctx != context.Background() {
		select {
		case <-ctx.Done():
			result.concludeCommand(CommandTag{}, newContextAlreadyDoneError(ctx))
			result.closed = true
			pgConn.unlock()
			return result
```

Which has a context check. You'd honestly have to check with Jack or run tests to be 100% sure, but it looks like it's checked.

I guess the way that the transaction gets rolled back then is implicitly, as the connection is killed the Postgres backend disposes of it automatically.